lib/charms/grafana_agent_k8s/v0/grafana_agent.py

```diff
@@ -155,37 +192,88 @@ def _on_log_proxy_relation_departed(self, event):
         self._update_agents_list(event)
 
         if len(self._current_config["clients"]) == 0:
-            self._container.stop(self._container_name)
+            self._container.stop("promtail")
         else:
-            self._container.restart(self._container_name)
+            self._container.restart("promtail")
 
     def _on_upgrade_charm(self, event):
         # TODO: Implement it ;-)
         pass
 
-    def _on_promtail_pebble_ready(self, _):
-        """Event handler for the pebble ready event.
+    def _get_container(self, container_name):
+        if container_name is not None:
+            return self._charm.unit.get_container(container_name)
 
-        Args:
-            event: The event object of the pebble ready event
-        """
+        containers = dict(self._charm.model.unit.containers)
+
+        if len(containers) == 1:
+            return self._charm.unit.get_container([*containers].pop())
+
+        # FIXME: Use custom exception
+        raise Exception("Container cannot be obtained")
 
+    def _build_pebble_layer(self):
+        """Event handler for the pebble ready event."""
         pebble_layer = {
             "summary": "promtail layer",
             "description": "pebble config layer for promtail",
             "services": {
                 "promtail": {
                     "override": "replace",
                     "summary": "promtail",
-                    "command": f"/usr/bin/promtail {self._cli_args}",
-                    "startup": "disable",
+                    "command": "{} {}".format(WORKLOAD_BINARY_PATH, self._cli_args),
+                    "startup": "enabled",
                 }
             },
         }
+        self._container.add_layer(self._container_name, pebble_layer, combine=True)
+
+    def _obtain_promtail(self, event) -> None:
+        if self._is_promtail_binary_in_workload():
+            return
+
+        self._download_promtail(event)
+
+        if not self._check_sha256sum():
+            msg = "Promtail bnary sha256sum mismatch"
+            logger.error(msg)
+            raise PromtailDigestError(msg)
+
+        self._unzip_binary()
+        self._upload_binary()
 
-        self._container.add_layer("promtail", pebble_layer, combine=True)
-        config = self._initial_config()
-        self._container.push(CONFIG_PATH, yaml.dump(config))
+    def _is_promtail_binary_in_workload(self) -> bool:
+        pat = WORKLOAD_BINARY_PATH.split("/")[-1]
+        return True if len(self._container.list_files(BINARY_DIR, pattern=pat)) == 1 else False
+
+    def _download_promtail(self, event) -> None:
+        url = json.loads(event.relation.data[event.unit].get("data"))["promtail_binary_zip_url"]
```

Comment:
Why is this URL not using the file constant?

Reply:
We are using the constant in the provider side where we are setting the URL in relation data:

```python
    def _on_log_proxy_relation_changed(self, event):
        if event.relation.data[self._charm.unit].get("data") is None:
            data = {}
            data.update(json.loads(self._loki_push_api))
            data.update(json.loads(self._promtail_binary_url))
            event.relation.data[self._charm.unit].update({"data": json.dumps(data)})
...

    @property
    def _promtail_binary_url(self) -> str:
        # FIXME: Use charmhub's URL
        return json.dumps({"promtail_binary_zip_url": PROMTAIL_BINARY_ZIP_URL})
```